src/factoid/responder.rs

```diff
@@ -0,0 +1,100 @@
+use super::util;
+use std::boxed::Box;
+use util::is_extension;
+
+pub struct Responders {
+    responders: Vec<Box<dyn Responder>>,
+}
+
+impl Responders {
+    pub fn new() -> Responders {
+        Responders {
+            responders: Vec::<Box<dyn Responder>>::new(),
+        }
+    }
+
+    pub fn register_responder<T: 'static + Responder>(&mut self, responder: T) {
+        self.responders.push(Box::new(responder));
+    }
+
+    pub fn respond(&self, input: &String) -> Vec<String> {
+        let mut outvec = Vec::new();
+        for resp in &self.responders {
+            if let Some(out) = resp.respond(input) {
```

Comment:
for the scala brain, this function is effectively: `self.responders.flat_map(|resp| resp.respond(input))`

which has a rust equivalent:
```
self.responders.filter_map(|resp| resp.respond(input)).collect()
```
(without `collect()` you'd have an unused iterator over the results)

Reply:
secondary note, you can avoid the `collect()` (and Vec allocation entirely!) by rewriting this function as `pub fn respond(&self, input: &String) -> impl Iter<Item=String>` and returning the iterator itself, i believe